code/game/machinery/doors/firedoor.dm

```diff
@@ -188,30 +182,51 @@
 		for(var/obj/machinery/firealarm/fire_panel in place.firealarms)
 			fire_panel.set_status()
 
+/obj/machinery/door/firedoor/proc/CalculateWatchedTurfs()
+	watched_turfs = list()
+	for(var/dir in GLOB.cardinals)
+		var/turf/checked_turf = get_step(get_turf(src),dir)
+		if(checked_turf)
+			watched_turfs |= checked_turf
+			RegisterSignal(checked_turf, COMSIG_TURF_EXPOSE, .proc/check_atmos)
+	watched_turfs |= get_turf(src)
+	check_atmos()
+
 /obj/machinery/door/firedoor/proc/check_atmos(datum/source)
+	SIGNAL_HANDLER
+	var/found_turf = FALSE
+
 	if(!COOLDOWN_FINISHED(src, detect_cooldown))
 		return
-	if(alarm_type)
+	if(alarm_type == FIRELOCK_ALARM_TYPE_GENERIC)
 		return
+	
 	for(var/area/place in affecting_areas)
 		if(!place.fire_detect) //if any area is set to disable detection
 			return
 
-	var/turf/my_turf = source
-	if(!my_turf)
-		my_turf = get_turf(src)
+	for(var/turf/checked_turf in watched_turfs)
+		if(source == checked_turf)
+			found_turf = TRUE
 
-	var/datum/gas_mixture/environment = my_turf.return_air()
 	var/result
 
-	if(environment?.temperature >= FIRE_MINIMUM_TEMPERATURE_TO_EXIST)
-		result = FIRELOCK_ALARM_TYPE_HOT
-	if(environment?.temperature <= BODYTEMP_COLD_DAMAGE_LIMIT)
-		result = FIRELOCK_ALARM_TYPE_COLD
-	if(!result)
-		return
+	for (var/turf/checked_turf in watched_turfs)
+		if(!checked_turf.density && found_turf)
+			var/datum/gas_mixture/environment = checked_turf.return_air()
+			if(environment?.temperature >= FIRE_MINIMUM_TEMPERATURE_TO_EXIST && my_area.active_alarms[ALARM_ATMOS])
+				result = FIRELOCK_ALARM_TYPE_HOT
+			if(environment?.temperature <= BODYTEMP_COLD_DAMAGE_LIMIT && my_area.active_alarms[ALARM_ATMOS])
+				result = FIRELOCK_ALARM_TYPE_COLD
```

Comment:
What happens if two firelocks are adjacent to each other?

Reply:
is this an open question? they wont select each other because its dense if closed